app/main.py

```diff
@@ -1,34 +1,149 @@
-        pass
+    def __init__(self,
+                 ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]) -> None:
+        self.ships = []
+        self.field = {(i, j): None for i in range(10) for j in range(10)}
+        for ship in ships:
+            ship = Ship(ship[0], ship[1])
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+        self._validate_field()
+
+        self.japan = Japan(self.field)
+        self.dolphin = Dolphin(self.field, self.japan.location)
+
+    def fire(self, location: Tuple[int, int]) -> str:
+        if location in self.field:
+            if self.field[location]:
+                status_before_fire = self.field[location].is_drowned
+                self.field[location].fire(*location)
+                status_after_fire = self.field[location].is_drowned
+                if status_before_fire == status_after_fire:
+                    return "Hit!"
+                return "Sunk!"
+            return "Miss!"
+
+    def print_field(self) -> None:
+        self.dolphin.swim(self.field)
+
+        print_field = ""
+        for location, cell in self.field.items():
+            if cell is None:
+                if location == self.japan.location:
+                    print_field += "\N{silhouette of japan}	"
+                elif location == self.dolphin.location:
+                    print_field += "\N{dolphin}	"
+                else:
+                    print_field += "\N{water wave}	"
+            elif cell.is_drowned:
+                print_field += "\N{anger symbol}	"
+            elif cell.get_deck(*location).is_alive:
+                print_field += "\N{sailboat}	"
+            else:
+                print_field += "\N{fire}	"
+
+            if location[1] == 9:
+                print_field += "\n"
+        print(print_field)
+
+    def _validate_field(self) -> None:
+        assert sorted(len(ship.decks) for ship in self.ships) == [
+            1, 1, 1, 1,
+            2, 2, 2,
+            3, 3,
+            4
+        ], ("There should be 10 ships: 4 single-deck ships, "
+            "3 double-deck ships, 2 three-deck ships and 1 four-deck ship")
+
+        for ship in self.ships:
+            for deck in ship.decks:
+                surroundings = get_surroundings((deck.row, deck.column))
+                for cell in surroundings:
+                    if cell in self.field:
+                        assert self.field[cell] in [None, ship], (
+                            "ships shouldn't be located "
+                            "in the neighboring cells"
+                        )
+
+
+class Japan:
+    """Used to spawn Japan in the `Battleship.print_field` method."""
+    def __init__(self, field_dict: Dict[Tuple[int, int], Ship | None]) -> None:
+        empty_cells = [cell for cell in field_dict if not field_dict[cell]]
+        self.location = random.choice(empty_cells)
+
+
+class Dolphin:
```

Comment:
What is this???

Reply:
![image](https://github.com/mate-academy/py-battleship/assets/134400101/e3384c42-61f9-4d83-b136-ef8fd9749468)

Its additional functionality, purely visual feature, doesn't change gameplay.